utils/docker/Dockerfile.centos.7

```diff
@@ -9,16 +9,72 @@
+ARG REPO_URL=""
+ARG REPO_DISTRO=""
+ARG REPO_DAOS=""
+
 # Intermittent cache-bust.  Used to reduce load on the actual CACHEBUST later.
 ARG CB0
 
+# If a local distro repository is supplied, then we should use it
+# and disable using the distro repositories.
+# The Docker image starts out with the DISTRO provided GPG keys
+# on the disk, but not installed.  These must be isntalled before
+# adding any repo that needs GPG keys but does not provide meta
+# data for installing those keys.  Then the epel-release
+# package must be installed and the GPG key it provides must be
+# installed.
+RUN if [ -n "$REPO_DISTRO" ]; then                              \
+      rpm --import /etc/pki/rpm-gpg/* ;                         \
+      MY_REPO="${REPO_URL}${REPO_DISTRO}/";                     \
+      MY_NAME="${MY_REPO#*//}";                                 \
+      MY_NAME="${MY_NAME//\//_}";                               \
+      echo -e "[${MY_NAME}]\n\
+name=created from ${MY_REPO}\n\
+baseurl=${MY_REPO}\n\
+enabled=1\n\
+repo_gpgcheck=0\n\
+gpgcheck=1\n" >> /etc/yum.repos.d/local-centos-group.repo;      \
+      yum -y install --disablerepo=extras                       \
+             dnf dnf-pluins-core epel-release-7-11;             \
+      rpm --import /etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7;         \
+      yum -y install dnf dnf-plugins-core;                      \
+      dnf config-manager --assumeyes --quiet                    \
+                         --disable base extras epel updates;    \
+      yum clean all;                                            \
+      dnf clean all;                                            \
+      sed "s/^mirrorlist_expire=0*/mirrorlist_expire=99999999/" \
+          -i /etc/yum.conf;                                     \
+    fi
+
+# If a local DAOS repository is supplied, then we should use it
+# This is mainly for quickbuilds
+RUN if [ -n "$REPO_DAOS" ]; then                                \
+      MY_REPO="${REPO_URL}${REPO_DAOS}/";                       \
+      MY_NAME="${MY_REPO#*//}";                                 \
+      MY_NAME="${MY_NAME//\//_}";                               \
+      echo -e "[${MY_NAME}]\n\
+name=created from ${MY_REPO}\n\
+baseurl=${MY_REPO}\n\
+enabled=1\n\
+repo_gpgcheck=0\n\
+exclude=fuse-3* fuse-devel-3* fuse-libs-3* fuse-debuginfo-3* ompi*\n\
+gpgcheck=0\n" >> /etc/yum.repos.d/local-daos-group.repo;        \
+    fi
+
 # Install OS updates and package.  Include basic tools and daos dependencies
 # that come from the core repo.
 # Clean up any repos afterwards to save space.
+# Switch to dnf as it seems a bit faster.
+# libatomic should be in this list, but can not for now due to CI
+# post provisioning issue.
+# *** Keep these in as much alphbetical order as possible ***
 RUN yum clean all && \
-    yum -y install epel-release && \
-    yum -y upgrade && \
-    yum -y install \
+    yum -y install dnf epel-release && \
```

Comment:
Earlier on in the REPO_DISTRO part a specific version of epel-release was requested but here it's not versioned.  It would be cleaner to just install these two packages on their own on L19 so it's common to all paths, although that might mean an extra metadata fetch.  We do need clarity on if we are concerned about which version of epel-release we get.

Additionally, the yum clean all on L73 can and should be removed.